### ISSUE TITLE
MCAPI Support for CMAF Media Profiles

### ISSUE THREAD

**Issue by johnsim** [WICG/media-capabilities#98]
ISO/IEC Common Media Application Format (CMAF) media profiles are identified by 4cc codes which are in the CMAF initialization segment and characterize the content. canPlayType and IsTypeSupported do not support the MIME type ‘profiles' sub-parameter, and I believe the same is true for MCAPI. 

The AV1 binding for CMAF can be fully characterized by the ‘codecs’ sub-parameter alone, but the same is not true for many other CMAF audio and video profiles. It would be desirable for MCAPI to be able to indicate support for CMAF media profiles. The media profile includes bit depth, chroma subsampling, color space, transfer functions, black level – which I believe cannot be expressed in the codecs string.

**Comment by chcunningham** [comment]
Thanks John. 
I follow you. Issue #67 proposes the same (though it started with a different title/agenda). I think its fine to move that discussion to its own bug (this one). FYI @wilaw @PauKerr 

I'll repaste my reply to Will here (the context of this reply is parsing profiles= as you suggest vs just adding additional fields to MC for the extra details like color, eotf, etc).

> I do have some concerns about profiles though.
> 
> Everytime a new profile is minted UAs will need to update their parsing code. As-is MC allows callers can supply any combination of values. This puts more burden on callers, but it affords them more flexibility and means deploying a new combination isn't blocked on UA releases.
> 
> Profiles constrain to a handful of colors and eotfs, but a handful is still ambiguous. With MC we aim to give confident answers (no more "maybe"), which motivates disallowing ambiguous inputs. We intentionally don't have a way to say we support some of the things in your profile. Perhaps an all-or-nothing approach could work, but it seems a bit fragile. The people that make devices and those that define profiles may not read my github posts ;).
> 
> Aside: for existing media APIs, it seems like additional prameters are by default ignored. For instance, the following returns "probably" in all edge, ff, and chrome.
> 
> document.createElement('video').canPlayType('video/mp4; codecs="avc1.42E01E"; profiles="foo"')
> 
> Not so much a problem for the shiny new MC api, but backward compat with other APIs is going be tricky.
> 


I should amend a part of that quote
> As-is MC allows callers can supply any combination of values.

This is only true if we add some extra fields to the API. We likely will add those irrespective of parsing profiles so that non-CMAF and non-AV1 streams can describe these things.

**Comment by cconcolato** [comment]
Note that a similar discussion is happening within MPEG, following are some bits of the discussion that are relevant. This calls for a common solution.

[RFC6381](https://tools.ietf.org/html/rfc6381#section-4.4) says:
> the 'profiles' parameter MUST list exactly the major-brand,
   followed by the compatible-brands, as listed in the filetype box
   ('ftyp') or segment-type box ('styp').  The major-brand MUST be
   first, and MAY be removed from the compatible-brands list.

But the ISOBMFF family of standards defines lots of brands, see [the list on the MP4 registration authority](http://mp4ra.org/#/brands), but not all brands (and therefore not all profiles) are defining media profiles, à la CMAF. 

@mikedo said:
> All brands are not created equal.  I think we have the following kinds of brands today:
> 1. File structure conformance, e.g. 'isob'.
> 2. Transfer encoding, e.g. encryption (and also maybe gzip) conformance.
> 3. Sample conformance, e.g. 'hev1'.
> 4. File structure + sample conformance, e.g. CMAF brands.
> 5. Supplemental data conformance, e.g. 'ccea'.

@dwsinger said:
> brands today:
> * express a claim (“file complies with the file requirements of the brand”)
> * express a permission (“if you recognize this brand and comply with the reader requirements, you can process the file and ignore what you don’t understand”)
> * are sometimes used with neither claims nor permissions, as a tag (“a Canon digital camera wrote this file”)
> * cannot express an “and” combination (“you need to comply with all of iso7,avc1,and cmaf to process this file, and the file complies with all of them”)

@irajs said:
> we had similar discussions for DASH profile signaling, not for replacing the "promise" and "permission" features of  brands, but provide an additional functionality for the parser/selector to signal the combined capabilities of the file/manifest.
> The 3 operations we looked at there, are:
> 1. AND: required to support all of
> 2. OR – i.e. adequate to support for any of
> 3. +: optionally support  of

**Comment by chcunningham** [comment]
Sorry for the delay. We've prioritized other issues, but this is still on my radar. I need to digest the brands from comment 2 - @wilaw how does that factor into your request for profiles? Look forward to discussion at FOMS.

**Comment by chcunningham** [comment]
@johnsim and I discussed in person. We both support the idea that MediaCapabilities can add fields for color matrix and transfer functions. Web authors could then completely describe any CMAF profile using a combination of the members from the MediaDecodingConfiguration.

I know this is not what @wilaw is hoping for and does not address the concern around the higher burden for callers of this api (using multiple fields instead of a single fourcc). I propose we address this concern via js library (which I'll happily kickstart) that maintains the mapping. This affords much faster updates and simplifies interop at the cost of having a thin layer between these users and MediaCapabilites. 

re: comments from cconcolato@, I don't personally have enough isobmff / cmaf expertise to contribute much to the reconciliation of brands/profiles between those specs.

**Comment by wilaw** [comment]
@chcunningham  - I support the compromise suggestion of a js polyfill approach. It will need to implement a look-up table between fourcc codes and the underlying core properties which MediaCapabilities exposes. Let me know once you kick-start and I'll get dash.js as an early implementer.

**Comment by chcunningham** [comment]
@wilaw in the recent WAVE call folks seemed to be moving towards putting more info into the manifest to make precise queries. Is the js polyfill work still relevant?

**Comment by wilaw** [comment]
I checked with Iraj this morning - MPEG has no plan to put lower-level attributes in to the manifest so for DASH side of the house it is status quo . On the HLS side I do not think Apple has any plans either to enrich playlists. So the polyfill approach still stands as a rational bridge between what the manifests will provide and the Media Capabilities API/dictionary surface.

**Comment by cconcolato** [comment]
FYI, [ISO/IEC 14496-12 6th edition FDAM 2](https://www.iso.org/standard/76596.html) ([N18418](http://wg11.sc29.org/doc_end_user/documents/126_Geneva/wg11/w18418.zip), for those who have access to internal MPEG docs) introduces a new tool which might be relevant to this discussion. 

The `profiles` sub-parameter exposes brands as typically declared in the `ftyp` box. However, the brands in the `ftyp` box are alternatives to each other. They indicate that if you understand at least one of them, you can do something with the file  (or-combination). In practice, this leads to an increase in the number of brands to be defined. If a new feature F is defined and becomes mandatory to interpret a file, but if that file is otherwise compliant to an existing brand B, you'd have to create a new version of that brand B' to say the file is compliant to B+F. The new tool is composed of 2 boxes similar to `ftyp` called `etyp` and `tyco`. They allow indicating an and-combination of brands. Citing the spec:

> Effectively, if the compatible_brands in the FileTypeBox are labeled C[1], C[2], etc., and the compatible_brands in the first TypeCombinationBox D[1,1], D[1,2], and the second D[2,1], D[2,2], and so on, then a reader has to support:
C[1] 
or C[2] 
or …
or ( D[1,1] and D[1,2] and …)
or ( D[2,1] and D[2,2] and …)
…

The RFC defining the `profiles` sub-parameters has not yet been updated for that, yet. Also, that does not solve the problem of mapping existing brands to Media Capabilities. But, ideally in the future, I think aligning brand definitions with Media Capabilities could enable creating less complex mapping implementations. I'm curious to hear if others share that view.

**Comment by dwsinger** [comment]
I don't agree that one updates _parsing_ code when there is either a new profile, or even a new MIME parameter. If a file says `profiles=CMAF,unkw,WLaw,Chris `and you recognize and support Chris (don't we all?), then you accept the file. It doesn't matter if unkw is there as well. You can parse this quite easily into the profiles parameter and the {CMAF,unkw,WLaw,Chris} set of values, without needing to know what the values _mean_.

If you write your code before the combo (as above) came out, and don't recognize `profiles=Chris,stuf+nons,Wlaw` again you should toss the stuff and nonsense.

Similarly, if a MIME parameter is unrecognized by you, why worry? Toss it. Maybe this is harder, as you have to know the general parsing rules to separate out parameters in general.

So, I think the spec. is backwards; rather than requiring the caller to strip unknown parameters, we should pass the whole lot to the media engine, and have it strip what it doesn't recognize.

**Comment by chrisn** [comment]
@chcunningham said:
> I propose we address this concern via js library (which I'll happily kickstart) that maintains the mapping.

@wilaw said: 
> I support the compromise suggestion of a js polyfill approach.

Also discussed in the 3 Nov 2020 MEIG meeting ([minutes](https://www.w3.org/2020/11/03-me-minutes.html)).

Has there been any progress on writing the JS library, and are there any remaining gaps for Media Capabilities?
